proof/invariant-abstract/AARCH64/ArchVCPUDomain_AI.thy

```diff
@@ -0,0 +1,597 @@
+  unfolding dissociate_vcpu_tcb_def
+  apply (wpsimp wp: hoare_vcg_imp_lift' get_vcpu_wp arch_thread_get_wp | wps)+
+  apply (clarsimp simp: cur_vcpu_def in_omonad split: option.splits)
+  done
+
+lemma associate_vcpu_tcb_valid_cur_vcpu[wp]:
+  "\<lbrace>cur_vcpu_in_cur_domain and ct_in_cur_domain and cur_vcpu and (\<lambda>s. tcb_ptr \<noteq> idle_thread s)
+    and (\<lambda>s. scheduler_action s = resume_cur_thread)\<rbrace>
+   associate_vcpu_tcb vcpu_ptr tcb_ptr
+   \<lbrace>\<lambda>_. cur_vcpu_in_cur_domain\<rbrace>" (is "\<lbrace>?P\<rbrace> _ \<lbrace>_\<rbrace>")
+  unfolding associate_vcpu_tcb_def
+  apply wpsimp
+        apply (solves \<open>(wpsimp wp: hoare_vcg_imp_lift' set_vcpu_not_current_cur_vcpu_in_cur_domain get_vcpu_wp
+                        | wps
+                        | clarsimp simp: in_omonad)+\<close>)+
+    apply (rule_tac Q'="\<lambda>_. ?P" in hoare_post_imp)
+     apply (fastforce simp: in_omonad ct_in_cur_domain_def cur_vcpu_def)
+    apply (wpsimp | wps)+
+  done
+
+lemma invoke_vcpu_ack_vppi_vcpu_tcbs_of[wp]:
+  "invoke_vcpu_ack_vppi vcpu_ptr vppi \<lbrace>\<lambda>s. P (vcpu_tcbs_of s)\<rbrace>"
+  unfolding invoke_vcpu_ack_vppi_def
+  apply wpsimp
+  by (fastforce simp: opt_map_def elim!: rsubst[where P=P])
+
+crunch invoke_vcpu_inject_irq, invoke_vcpu_write_register, invoke_vcpu_ack_vppi
+  for vcpu_tcbs_of[wp]: "\<lambda>s. P (vcpu_tcbs_of s)"
+  and arm_current_vcpu[wp]: "\<lambda>s. P (arm_current_vcpu (arch_state s))"
+  and cur_vcpu_in_cur_domain[wp]: cur_vcpu_in_cur_domain
+  (wp: cur_vcpu_in_cur_domain_lift_strong)
+
+crunch invoke_vcpu_read_register
+  for cur_vcpu_in_cur_domain[wp]: cur_vcpu_in_cur_domain
+
+lemma perform_vcpu_invocation_valid_cur_vcpu[wp]:
+  "\<lbrace>cur_vcpu_in_cur_domain and ct_in_cur_domain and cur_vcpu and valid_vcpu_invocation vi
+    and (\<lambda>s. scheduler_action s = resume_cur_thread)\<rbrace>
+   perform_vcpu_invocation vi
+   \<lbrace>\<lambda>_. cur_vcpu_in_cur_domain\<rbrace>"
+  unfolding perform_vcpu_invocation_def
+  by (wpsimp simp: valid_vcpu_invocation_def)
+
+crunch store_pte, store_asid_pool_entry
+  for cur_vcpu_in_cur_domain[wp]: cur_vcpu_in_cur_domain
+  (wp: cur_vcpu_in_cur_domain_lift)
+
+crunch
+  perform_vspace_invocation, perform_page_table_invocation, perform_page_invocation,
+  perform_asid_control_invocation, perform_asid_pool_invocation, perform_sgi_invocation
+  for cur_vcpu_in_cur_domain[wp]: cur_vcpu_in_cur_domain
+  (wp: crunch_wps  simp: crunch_simps)
+
+lemma arch_perform_invocation_valid_cur_vcpu[wp]:
+  "\<lbrace>cur_vcpu_in_cur_domain and ct_in_cur_domain and cur_vcpu and valid_arch_inv ai
+    and (\<lambda>s. scheduler_action s = resume_cur_thread)\<rbrace>
+   arch_perform_invocation ai
+   \<lbrace>\<lambda>_. cur_vcpu_in_cur_domain\<rbrace>"
+  unfolding arch_perform_invocation_def
+  by (wpsimp simp: valid_arch_inv_def)
+
+\<comment> \<open>FIXME: move (where?)\<close>
```

Comment:
bit confused about relationship between `ArchVCPUDomain_AI` and `ArchVCPU_AI`... possibly I'd expect it to go into the latter?

Reply:
same for the other `(where?)` above in the file

Reply:
I guess the other option would be moving them into `IpcCancel_AI` next to the equivalent lemmas for bound TCBs and notifications. `TcbAcc_AI` does feel a bit more suitable though, so I might move all four of these there instead.